### ISSUE TITLE
AD: custom-VJP gradient (e.g. vqe-energy) silently returns zeros inside a nested define

### ISSUE THREAD

**Issue by RichardHoekstra** [tsotchke/eshkol#296]
## Summary

`gradient` through a function with a custom vector-Jacobian-product (e.g. `vqe-energy` from `agent.quantum`) **silently returns a zero vector** when the `gradient` call sits inside a nested `define`d helper, but returns the correct gradient when the same call is written in the enclosing `lambda`/scope.

It fails **silently** (zeros, no error), which is the dangerous part: a perfectly ordinary training loop written with helper functions compiles, runs, and simply never learns — the energy/loss stays frozen with no diagnostic.

Pure-Eshkol-function gradients are **not** affected in the same nested scope, so this is specific to the custom-VJP / external-adjoint path, not nested-scope AD in general.

## Environment

- Eshkol Compiler **v1.3.3-evolve** (commit `831894f`)
- Built with `-DESHKOL_QUANTUM_ENABLED=ON`, LLVM 21.1.8, Ubuntu 22.04
- Run via `eshkol-run -r`

## Minimal repro

```scheme
(require agent.quantum)

;; A pure Eshkol function, differentiated inside a nested define (works fine).
(define (pure-sq v) (let ((a (vector-ref v 0))) (* a a)))
(define (nested-pure-grad v) (gradient pure-sq v))

(with-hamiltonian (make-h2-hamiltonian 0.7414)
  (lambda (ham)
    (define p (vector 0.1 0.2 0.3 0.4))

    ;; (A) custom-VJP gradient in the enclosing lambda scope
    (display "A. vqe grad, enclosing scope : ")
    (display (gradient (lambda (q) (vqe-energy ham q)) p)) (newline)

    ;; (B) the SAME call inside a nested define
    (define (nested-vqe-grad v) (gradient (lambda (q) (vqe-energy ham q)) v))
    (display "B. vqe grad, nested define   : ")
    (display (nested-vqe-grad p)) (newline)

    ;; (C) control: a PURE-function gradient inside a nested define
    (display "C. pure grad, nested define  : ")
    (display (nested-pure-grad p)) (newline)))
```

## Actual output

```
A. vqe grad, enclosing scope : #(0.254119 -0.00358856 0.113678 -2.11092e-18)
B. vqe grad, nested define   : #(0 0 0 0)
C. pure grad, nested define  : #(0.2 0 0 0)
```

- **A** is correct (bit-identical to Moonlab's `vqe-gradient` adjoint at these params).
- **B** is the bug: the *same* gradient expression returns all zeros purely because it is inside a nested `define`.
- **C** shows a pure-function gradient in the *same* nested scope is fine (`d/dx x² = 2x = 0.2` at `x=0.1`), which isolates the problem to the custom-VJP path.

## Expected

**B** should equal **A** — a gradient must not depend on the lexical nesting of the call site. Failing that, it should raise an error rather than silently return zeros.

## Impact

This is easy to hit and hard to notice. The natural way to write a VQE/optimization loop is with a `gradient-step`/`train` helper (as `examples/gradient_descent_demo.esk` does for pure functions); doing the same with `vqe-energy` produces a loop that runs cleanly but never moves. The workaround is to inline the `gradient` call into the enclosing scope, but that isn't discoverable without diffing against a working reference.

**Comment by tsotchke** [comment]
Confirmed and triaged at the highest severity we track — a silently wrong gradient is the worst failure class this compiler can have, and your isolation (pure-function AD fine in the same scope, custom-VJP path not) saved us real time. Thank you for the minimal repro.

Status: reproducing now on a quantum-enabled build of master with two added discriminators (Hamiltonian passed as a parameter instead of captured; let-bound helper instead of nested define) to pin down whether this is the reverse/forward mode-selection analysis failing to see through the nested-define lambda lifting, or transitive capture resolution on the custom-VJP path. The fix will land with a regression test covering exactly your A/B/C shape, and — independent of the root cause — the failure mode will be made loud: if a custom-VJP node can't participate in a gradient, that will be an error, never a zero vector.

Will link the PR here.
